File: AmaroTest/app/src/main/java/com/test/amaro/amarotest/Views/ProductsList/ProductsListInteractor.java
package com.test.amaro.amarotest.Views.ProductsList;

import android.content.Context;
import android.util.Log;

import com.android.volley.Request;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.JsonObjectRequest;
import com.test.amaro.amarotest.Model.Product;
import com.test.amaro.amarotest.WebServices.ConfigURL;
import com.test.amaro.amarotest.WebServices.VolleySingleton;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.List;

/**
 * Created by jaiber on 1/23/18.
 */

public class ProductsListInteractor {

    private static final String TAG = "ProductsListInteractor" ;
    private List<Product> productList;
    private Context ctx;
    private OnDataRequestListener onDataRequestListener;

    public interface OnDataRequestListener {

        void onDataReady(List<Product> productList);

    }

    public ProductsListInteractor(Context _ctx, OnDataRequestListener listener){
        ctx = _ctx;
        onDataRequestListener = listener;
    }

    public void requestProducts() {

        Log.d(TAG, "request");

        JsonObjectRequest jsonObjReq = new JsonObjectRequest(
                Request.Method.GET,
                ConfigURL.URL_PRODUCTS,
                null,
                new Response.Listener<JSONObject>() {

                    @Override
                    public void onResponse(JSONObject response) {
                        //Log.d(TAG, response.toString());
                        try {
                            JSONArray productsJson = response.getJSONArray("products");
                            productList = Product.fromJson(productsJson);

                            if(productList != null) {
                                for(int i=0; i<productList.size(); i++){
                                    Log.d("*Products", "*Product Name: " + productList.get(i).getName());
                                    Log.d("*Products", "*Price: " + productList.get(i).getActualPrice());
                                }
                                onDataRequestListener.onDataReady(productList);
                            }else {
                                Log.d(TAG, "data is null ");
                            }

                        } catch (JSONException e) {
                            e.printStackTrace();
                        }


                    }
                }, new Response.ErrorListener() {

            @Override
            public void onErrorResponse(VolleyError error) {

            }
        });
        VolleySingleton.getInstance(ctx).addToRequestQueue(jsonObjReq);
    }

}

File: AmaroTest/app/src/main/java/com/test/amaro/amarotest/Views/ProductsList/ProductsListInteractor.kt
package com.test.amaro.amarotest.Views.ProductsList

import android.content.Context
import android.util.Log
import com.test.amaro.amarotest.model.Product
import com.test.amaro.amarotest.api.ApiService
import com.test.amaro.amarotest.model.Products
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import timber.log.Timber

/**
 * Created by jaiber on 1/23/18.
 */

class ProductsListInteractor(private val ctx: Context, private val onDataRequestListener: OnDataRequestListener, private val api: ApiService) {

    private var productList: ArrayList<Product>? = null
    //internal var `var`: lateinit? = null


    interface OnDataRequestListener {

        fun onDataReady(productList: List<Product>?)

    }

    companion object {

        private val TAG = "ProductsListInteractor"
    }

    fun requestProducts(){

        Timber.i("on requestProducts")
        api.requestProducts().enqueue(object : Callback<Products> {
            override fun onResponse(call: Call<Products>?, response: Response<Products>) {
                Timber.i(response.toString())
                if(response.isSuccessful){
                    productList = response.body()?.products
                    Log.d("TAG", ""+response.toString())

                    if(productList != null)
                    onDataRequestListener.onDataReady(productList)
                }
            }

            override fun onFailure(call: Call<Products>, t: Throwable) {
                Timber.i("errorrrr ${t.message}")
            }
        })
    }

  /*  fun requestProducts() {

        Log.d(TAG, "request")

        val jsonObjReq = JsonObjectRequest(
                Request.Method.GET,
                ConfigURL.INSTANCE.getURL_PRODUCTS(), null,
                Response.Listener { response ->
                    //Log.d(TAG, response.toString());
                    try {
                        val productsJson = response.getJSONArray("products")
                        productList = Product.Companion.fromJson(productsJson)

                        if (productList != null) {
                            for (i in productList!!.indices) {
                                Log.d("*Products", "*Product Name: " + productList!![i].name!!)
                                Log.d("*Products", "*Price: " + productList!![i].actualPrice!!)
                            }
                            onDataRequestListener.onDataReady(productList)
                        } else {
                            Log.d(TAG, "data is null ")
                        }

                    } catch (e: JSONException) {
                        e.printStackTrace()
                    }
                }, Response.ErrorListener { })
        VolleySingleton.getInstance(ctx).addToRequestQueue(jsonObjReq)
    }
*/


}
